### DISCUSSION TITLE
Could this repo utilize `pnpm` for restoring node packages?

### DISCUSSION THREAD

**Discussion by sgalcheung** [edrlab/thorium-reader#2133] [Developer corner]
First, I use `pnpm install` to restore the node packages in this repository.
Then, I run the build with `npm run build`, but I encountered some syntax issues.
> EDRLab.ThoriumReader@3.0.0-alpha.1 build
> cross-env NODE_ENV=production webpack --config webpack.config.js && ncp src/package.json dist/package.json
> PREPROCESSOR nodeEnv: production
> MAIN nodeEnv: production
> WEBPACK externals (MAIN): --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
> {
>   "bindings": "bindings",
>   "fsevents": "fsevents",
>   "electron-devtools-installer": "electron-devtools-installer",
>   "remote-redux-devtools": "remote-redux-devtools",
>   "electron": "electron",
>   "yargs": "yargs"
> }
> LIBRARY nodeEnv: production
> WEBPACK externals (LIBRARY): --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
> {
>   "bindings": "bindings",
>   "fsevents": "fsevents",
>   "electron-devtools-installer": "electron-devtools-installer",
>   "remote-redux-devtools": "remote-redux-devtools",
>   "electron": "electron",
>   "yargs": "yargs"
> }
> READER nodeEnv: production
> WEBPACK externals (READER): --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
> {
>   "bindings": "bindings",
>   "fsevents": "fsevents",
>   "electron-devtools-installer": "electron-devtools-installer",
>   "remote-redux-devtools": "remote-redux-devtools",
>   "electron": "electron",
>   "yargs": "yargs"
> }
> PDF nodeEnv: production
> WEBPACK externals (PDF): --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
> {
>   "bindings": "bindings",
>   "fsevents": "fsevents",
>   "electron-devtools-installer": "electron-devtools-installer",
>   "remote-redux-devtools": "remote-redux-devtools",
>   "electron": "electron",
>   "yargs": "yargs"
> }
> PRELOAD nodeEnv: production
> [webpack-cli] ModuleNotFoundError: Module not found: Error: Can't resolve 'urijs' in '/Users/
> ......
>     at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
>   errno: -32,
>   code: 'EPIPE',
>   syscall: 'write'
> }
> Error: write EPIPE
>     at target._send (node:internal/child_process:879:20)
>     at target.send (node:internal/child_process:752:19)
> ......
>     at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
>   errno: -32,
>   code: 'EPIPE',
>   syscall: 'write'

}

**Comment by sgalcheung** [comment]
I'll walk through the steps to resolve this issue, summarizing the need to add a package.
```
pnpm add urijs --save-dev
pnpm add tslib --save-dev
pnpm add moment
pnpm add slugify --save-dev
pnpm add tough-cookie
pnpm add @types/tough-cookie --save-dev
pnpm add @redux-saga/core
pnpm add debug --save-dev
pnpm add @types/debug --save-dev
pnpm add bindings --save-dev
```

**Reply by danielweck** [reply]
the cookie package (other than types) is a runtime dependency. you are incorrectly adding the packages as development dependencies.
probably other packages too (moment, Redux Saga...)
I would advise against trying to hack packages into a fresh PNPM config, better spend time trying to understand why the package manager errors occurred in the first place.

or better even ... just use NPM :)

**Reply by sgalcheung** [reply]
thank you for your reply, I update it.

**Comment by danielweck** [comment]
Hello, this code repository's software dependencies (including postinstall scripts policy, peer dependencies strategy, semantic versioning resolutions, etc.) are handled by the official NodeJS package manager (NPM).
This project does not officially support PNPM, or Yarn, or Bun etc.
The package "lock" file defines the exact versions that the software is tested with. Sometimes even non-major semver deviations (i.e. minor or patch) can introduce significant changes in Thorium's build process or runtime due to lower-quality transitive dependencies in the Node ecosystem (or just human error, to be fair).
We are a very small development team so we cannot offer support for any other build environment / dependency graph, sorry. You might want to replicate Thorium's package-lock.json to match your preferred package manager, in order to mirror Thorium's currently-tested environment.
I am moving this issue to a GitHub discussion as this is an interesting topic, but this is not actionable (not supported, no fix, no resolution).

**Comment by sgalcheung** [comment]
After adding these packages, the project can run normally and no side effects have been seen so far.

**Reply by danielweck** [reply]
you will definitely risk problems during production builds, which do not include development dependencies :)

**Reply by sgalcheung** [reply]
Thank you for your reply. At present, I mainly want to test whether my lcp service is normal and can read lcpl files.

**Reply by danielweck** [reply]
FYI the "test" / alpha builds of Thorium available here in GitHub (compiled from the develop branch automatically) do not support LCP, unlike the official releases (latest is v2.4.1).
Do you have the LCP native library to test protected publications?

**Reply by sgalcheung** [reply]
I'm new to Readium LCP, my situation is that I want to deploy a similar system for knowledge payment. After some troubles, I started the [lsp server](https://github.com/readium/readium-lcp-server) and obtained the lcpl file. I tested it on the reader side to see if it is normal.

> The import of /Users/sgalcheung/Downloads/first-epub-book-sgal-cheung.lcpl failed. Error: end of central directory record signature not found

So I want to debug to figure out where the problem is

**Reply by danielweck** [reply]
your LCPL license JSON should contain a "publication" link that points to a downloadable EPUB file from your test Go server. The error message indicates that the downloaded file is not a  valid zip archive.

**Reply by sgalcheung** [reply]
Oh yes, I realized it by reading the [lsp architecture](https://github.com/readium/readium-lcp-server/wiki).
Thank you for mentioning this, I'll check my Go server again.
My server manage the storage of encrypted publications is local filesystem, which config by 
storage:>filesystem:>directory  in config.yaml
 I found this file store in right place, that is no extension, not zip file. Is this what's causing my problem?

> -rw-r--r--  1 sgalcheung  admin    48K May  3 16:42 fd53863e-6e7e-4c30-8f95-f0eb6b73c678

**Reply by sgalcheung** [reply]
I found this download url (this file alice.lcpl, download from [testing LCP / LSD server](https://[front-prod.edrlab.org/frontend/purchases/14/status](https://front-prod.edrlab.org/frontend/purchases/14/status)))
![image](https://github.com/edrlab/thorium-reader/assets/42566429/41a479b3-4f84-4b41-9f5e-e6254fccb397)
this type is `"application/epub+zip"`
and my test lcpl file is `"application/vnd.readium.license.status.v1.0+json"`
![image](https://github.com/edrlab/thorium-reader/assets/42566429/c252e4e2-016c-40c7-b2d9-8e7180d6a8ab)

it is the problem?

**Reply by danielweck** [reply]
"status" is for LSD JSON, not LCP JSON (different file types)

**Reply by sgalcheung** [reply]
What's wrong? I found this URL configured the `config.yaml` under the [license>links>status](https://github.com/readium/readium-lcp-server/blob/6e95190e99f175708df76ec7685f4442c2179a7d/test/config.yaml#L51) section , and the comment says"leave the url as-is (after <LSD_SERVER> has been resolved)". So I configured this to LSD server host. Shouldn't it be configured to the LCP server instead? Also, I noticed there's another section configured for [publication](https://github.com/edrlab/thorium-reader/discussions/2133#discussioncomment-9313488)(This is what the above mentioned should generate into the lcpl file containing the publication?), but this is [deprecated](https://github.com/readium/readium-lcp-server#license-section). What happened?

**Reply by sgalcheung** [reply]
after adding the publication section:
`publication: "http://192.168.50.12:8989/contents/{publication_id}"`
it works well. I made a mistake, but I also suggest adding more details to the `config.yaml` for this. Thanks!

**Comment by sgalcheung** [comment]
> FYI the "test" / alpha builds of Thorium available here in GitHub (compiled from the develop branch automatically) do not support LCP, unlike the official releases (latest is v2.4.1). Do you have the LCP native library to test protected publications?

do you means like this [lcp-testing-tools](https://github.com/edrlab/lcp-testing-tools) ?
